### PR TITLE
Overflow

### LINKED ISSUE #878: Incorrect result ?
HI guys, 
Consider the following program:

```
input relation Z(a:string)
relation PQRI(a:string)
relation PLEY(a:string)
relation NFUV(a:string)
output relation OUT(a:string)



PQRI(v) :- Z(v), Z(nbj).
PLEY(o) :- PQRI(x), Z(o), Z(x).
NFUV(q) :- Z(fym), PLEY(q).
OUT(t) :- NFUV(ssz), PLEY(arv), PLEY(t).
```
If I run the program (after compiling in release mode), I get:

```
$ ./rules_orig_cli < facts.dat 
OUT{.a = "3u37ezDdBg"}
OUT{.a = "EbgMd2DV3K"}
OUT{.a = "uwsUc"}
OUT{.a = "zDdBgEbgMd"}
```

Now if i just add `PQRI(z), PQRI(x)` in rule `PLEY` to get a new program:

```
input relation Z(a:string)
relation PQRI(a:string)
relation PLEY(a:string)
relation NFUV(a:string)
output relation OUT(a:string)



PQRI(v) :- Z(v), Z(nbj).
PLEY(o) :- PQRI(x), Z(o), Z(x), PQRI(z), PQRI(x). // modified rule
NFUV(q) :- Z(fym), PLEY(q).
OUT(t) :- NFUV(ssz), PLEY(arv), PLEY(t).
```

after recompiling and running it again, now I get an empty result.

```
$ ./rules_orig_cli < facts.dat 

```
However, according to my understanding, this modification should not change the result of the computation. 
Right? or am I misunderstanding something?

```
$ cat facts.dat 
start;


insert Z("3u37ezDdBg"),
insert Z("uwsUc"),
insert Z("zDdBgEbgMd"),
insert Z("EbgMd2DV3K");

commit;

dump OUT;

```
I am running this on ddlog version: 5b1c9c1a70992032d271934f2652dcf6952b390d
Operating System: `Debian GNU/Linux 10 (buster)`
Kernel: `Linux 4.19.75.1.amd64-smp`
Architecture: `x86-64`

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
Feature to panic on weight overflow.
This partly fixes #878: instead of giving an incorrect result the program will crash.
Still, the user has no (simple) recourse.

### CODE REVIEW

File: rust/template/differential_datalog/src/program/mod.rs [hunk 1]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+    }
+}
+
+impl Neg for CheckedWeight {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        // intentional panic on overflow
+        CheckedWeight {
+            value: 0_i32.checked_sub(self.value).expect("Weight overflow"),
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:147]: This is redundant, negation can't fail since the positive and negative integer spaces are equal. Any positive number can be converted into a negative one

File: rust/template/differential_datalog/src/program/mod.rs [hunk 2]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
 type TValEnter<P, T> = TraceEnter<TValAgent<P>, T>;
 type TKeyEnter<P, T> = TraceEnter<TKeyAgent<P>, T>;
 
-/// Diff associated with records in differential dataflow
+#[derive(
+    Abomonation, Copy, Ord, PartialOrd, Eq, PartialEq, Hash, Debug, Clone, Serialize, Deserialize,
+)]
+pub struct CheckedWeight {
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:95]:
```suggestion
#[derive(
    Abomonation, Copy, Ord, PartialOrd, Eq, PartialEq, Hash, Debug, Clone, Serialize, Deserialize,
)]
#[repr(transparent)]
pub struct CheckedWeight {
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 3]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+    pub value: i32,
+}
+
+/// CheckedWeight trait implementations
+
+impl Semigroup for CheckedWeight {
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:101]:
```suggestion
impl Semigroup for CheckedWeight {
```
This doc comment doesn't add anything

File: rust/template/differential_datalog/src/program/mod.rs [hunk 4]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+            .expect("Weight overflow");
+    }
+}
+
+impl Add for CheckedWeight {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self {
+        // intentional panic on overflow
+        CheckedWeight {
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:122]:
```suggestion
        Self {
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 5]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+                .expect("Weight overflow"),
+        }
+    }
+}
+
+impl Mul for CheckedWeight {
+    type Output = Self;
+    fn mul(self, rhs: Self) -> Self::Output {
+        // intentional panic on overflow
+        CheckedWeight {
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:135]:
```suggestion
        Self {
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 6]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+        }
+    }
+}
+
+impl Neg for CheckedWeight {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        // intentional panic on overflow
+        CheckedWeight {
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:146]:
```suggestion
        Self {
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 7]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+            value: 0_i32.checked_sub(self.value).expect("Weight overflow"),
+        }
+    }
+}
+
+impl Monoid for CheckedWeight {
+    fn zero() -> Self {
+        CheckedWeight { value: 0 }
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:154]:
```suggestion
        Self { value: 0 }
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 8]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+    }
+}
+
+impl One for CheckedWeight {
+    fn one() -> Self {
+        CheckedWeight { value: 1 }
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:160]:
```suggestion
        Self { value: 1 }
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 9]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+    }
+}
+
+impl Zero for CheckedWeight {
+    fn zero() -> Self {
+        CheckedWeight { value: 0 }
+    }
+    fn is_zero(&self) -> bool {
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:168]:
```suggestion
        Self { value: 0 }
    }

    fn is_zero(&self) -> bool {
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 10]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+        self.value == 0
+    }
+}
+
+impl From<i32> for CheckedWeight {
+    fn from(item: i32) -> Self {
+        CheckedWeight { value: item }
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:175]:
```suggestion
    fn from(value: i32) -> Self {
        Self { value }
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 11]
```diff
@@ -85,7 +89,111 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
...
+    }
+}
+
+impl From<i8> for CheckedWeight {
+    fn from(item: i8) -> Self {
+        CheckedWeight { value: item as i32 }
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:181]:
```suggestion
        Self { value: item as i32 }
```

File: rust/template/differential_datalog/src/program/worker.rs [hunk 12]
```diff
@@ -66,7 +67,13 @@ struct SessionData {
...
         ArrId,
         TraceAgent<
-            Spine<DDValue, DDValue, u32, i32, Rc<OrdValBatch<DDValue, DDValue, u32, i32, u32>>>,
+            Spine<
+                DDValue,
+                DDValue,
+                u32,
+                Weight,
+                Rc<OrdValBatch<DDValue, DDValue, u32, Weight, u32>>,
+            >,
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/worker.rs:76]: I believe this can be replaced with `OrdValSpine`

File: rust/template/differential_datalog/src/program/mod.rs [hunk 13]
```diff
@@ -1898,7 +2006,7 @@ impl Program {
                                 &collection_keys,
                                 arr.filter(move |_, v| f(v)),
                                 |k, key| *key = k.clone(),
-                                move |_, w1, v2, w2| (jfun(v2), w1 * w2),
+                                move |_, w1, v2, w2| (jfun(v2), (*w1) * (*w2)),
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:2009]:
```suggestion
                                move |_, &w1, v2, &w2| (jfun(v2), w1 * w2),
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 14]
```diff
@@ -1887,7 +1995,7 @@ impl Program {
                                 &collection_keys,
                                 arr.clone(),
                                 |k, key| *key = k.clone(),
-                                move |_, w1, v2, w2| (jfun(v2), w1 * w2),
+                                move |_, w1, v2, w2| (jfun(v2), (*w1) * (*w2)),
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:1998]:
```suggestion
                                move |_, &w1, v2, &w2| (jfun(v2), w1 * w2),
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 15]
```diff
@@ -1847,7 +1955,7 @@ impl Program {
                                 &collection_with_keys,
                                 arr.filter(move |_, v| f(v)),
                                 |(k, _), key| *key = k.clone(),
-                                move |v1, w1, v2, w2| (jfun(v2, &v1.1), w1 * w2),
+                                move |v1, w1, v2, w2| (jfun(v2, &v1.1), (*w1) * (*w2)),
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:1958]:
```suggestion
                                move |v1, &w1, v2, &w2| (jfun(v2, &v1.1), w1 * w2),
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 16]
```diff
@@ -1836,7 +1944,7 @@ impl Program {
                                 &collection_with_keys,
                                 arr.clone(),
                                 |(k, _), key| *key = k.clone(),
-                                move |v1, w1, v2, w2| (jfun(v2, &v1.1), w1 * w2),
+                                move |v1, w1, v2, w2| (jfun(v2, &v1.1), (*w1) * (*w2)),
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:1947]:
```suggestion
                                move |v1, &w1, v2, &w2| (jfun(v2, &v1.1), w1 * w2),
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 17]
```diff
@@ -1628,7 +1736,7 @@ impl Program {
                         &collection_with_keys,
                         arr,
                         |(k, _), key| *key = k.clone(),
-                        move |v1, w1, _, w2| (jfun(&v1.1), w1 * w2),
+                        move |v1, w1, _, w2| (jfun(&v1.1), (*w1) * (*w2)),
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:1739]:
```suggestion
                        move |v1, &w1, _, &w2| (jfun(&v1.1), w1 * w2),
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 18]
```diff
@@ -1595,7 +1703,7 @@ impl Program {
                         &collection_with_keys,
                         arr,
                         |(k, _), key| *key = k.clone(),
-                        move |v1, w1, v2, w2| (jfun(&v1.1, v2), w1 * w2),
+                        move |v1, w1, v2, w2| (jfun(&v1.1, v2), (*w1) * (*w2)),
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:1706]:
```suggestion
                        move |v1, &w1, v2, &w2| (jfun(&v1.1, v2), w1 * w2),
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 19]
```diff
@@ -1402,7 +1510,7 @@ impl Program {
                         &collection_with_keys,
                         arr,
                         |(k, _), key| *key = k.clone(),
-                        move |v1, w1, _, w2| (jfun(&v1.1), w1 * w2),
+                        move |v1, w1, _, w2| (jfun(&v1.1), (*w1) * (*w2)),
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:1513]:
```suggestion
                        move |v1, &w1, _, &w2| (jfun(&v1.1), w1 * w2),
```

File: rust/template/differential_datalog/src/program/mod.rs [hunk 20]
```diff
@@ -1369,7 +1477,7 @@ impl Program {
                         &collection_with_keys,
                         arr,
                         |(k, _), key| *key = k.clone(),
-                        move |v1, w1, v2, w2| (jfun(&v1.1, v2), w1 * w2),
+                        move |v1, w1, v2, w2| (jfun(&v1.1, v2), (*w1) * (*w2)),
```
Comment (Kixiron) [inline rust/template/differential_datalog/src/program/mod.rs:1480]:
```suggestion
                        move |v1, &w1, v2, &w2| (jfun(&v1.1, v2), w1 * w2),
```

**Review by Kixiron — CHANGES_REQUESTED** [review]
There's also an extraneous file, `test/datalog_tests/prof.profile`

**Comment by ryzhyk** [comment]
@mbudiu-vmw, can you please measure how this feature affects the benchmarks in `differential-datalog/rust/ddlog_benches/`?

**Comment by mbudiu-vmw** [comment]
I will measure perf, but we should also run all  other programs in this mode. We could also add configs for bigger weights, although even 64 bits is not much.

**Comment by mbudiu-vmw** [comment]
The best thing about this patch is that it immediately suggests a safe solution which I plan to prototype: make the weight an `Either<int32, BigInt>`.

**Comment by ryzhyk** [comment]
> I will measure perf, but we should also run all other programs in this mode. We could also add configs for bigger weights, although even 64 bits is not much.

This PR should configure a subset of the test suite to run with this feature enabled.

File: rust/template/differential_datalog/src/program/mod.rs (hunk 1 above)
Comment (mbudiu-vmw) [reply rust/template/differential_datalog/src/program/mod.rs:147]: They are not really equal, but indeed the negative one is bigger than the positive one.

Comment (mbudiu-vmw) [reply rust/template/differential_datalog/src/program/mod.rs:147]: But this check is actually necessary, in case that self.value is MIN_INT: https://towardsdatascience.com/unsinged-signed-integers-and-casting-in-rust-9a847bfc398f#3e54

File: rust/template/differential_datalog/src/program/mod.rs (hunk 2 above)
Comment (mbudiu-vmw) [reply rust/template/differential_datalog/src/program/mod.rs:95]: > 
> This PR should configure a subset of the test suite to run with this feature enabled.

Any suggestions? Do we have any tools to measure the perf difference?

Comment (ryzhyk) [reply rust/template/differential_datalog/src/program/mod.rs:95]: This is just for correctness, not performance.  Speaking of which, a test that actually overflows and crashes would be nice too.

As far as performance goes, the benchmarks in `differential-datalog/rust/ddlog_benches` is the best we have right now.  (see my other comment).

Comment (mbudiu-vmw) [reply rust/template/differential_datalog/src/program/mod.rs:95]: The test overflow.dl crashes with this feature. It is based on the bug report in #878

**Comment by mbudiu-vmw** [comment]
On the twitter-micro benchmark the perf impact is in the noise: most tests have unchanged perf, some are faster, and some are slower (according to cargo bench).

**Comment by mbudiu-vmw** [comment]
twitter-macro benchmark results are similar. This is measured in a VM, so there is significant noise.

File: rust/ddlog_benches/Makefile.toml [hunk 21]
```diff
@@ -1,6 +1,6 @@
 # Runs all ddlog benchmarks
 [tasks.benchmarks]
-dependencies = ["download-data", "bench-twitter", "bench-livejournal"]
+dependencies = ["bench-twitter", "bench-livejournal"]
```
Comment (Kixiron) [inline rust/ddlog_benches/Makefile.toml:3]: What's the reason for this change?

File: .gitignore [hunk 22]
```diff
@@ -31,6 +31,7 @@ test/datalog_tests/*.dump
 test/datalog_tests/*.c
 test/datalog_tests/*_ddlog
 test/datalog_tests/debug.log
+test/datalog_tests/prof.profile
```
Comment (Kixiron) [inline .gitignore:34]: This should go in `.git/info/exclude` instead of the committed gitignore

File: rust/ddlog_benches/Makefile.toml (hunk 21 above)
Comment (mbudiu-vmw) [reply rust/ddlog_benches/Makefile.toml:3]: without this change 'cargo bench' fails.

Comment (Kixiron) [reply rust/ddlog_benches/Makefile.toml:3]: What's the error message?

File: .gitignore (hunk 22 above)
Comment (mbudiu-vmw) [reply .gitignore:34]:
```
[cargo-make] INFO - Running Task: download-data
downloading '' to 'data/' 
[cargo-make] ERROR - Error while running duckscript: Source: Unknown Line: 10 - Invalid URL:  provided.
[cargo-make] WARN - Build Failed.
```
There isn't really a task download-data, it is a task invoked by other tasks.

File: rust/ddlog_benches/Makefile.toml (hunk 21 above)
Comment (Kixiron) [inline rust/ddlog_benches/Makefile.toml:3]:
```suggestion
dependencies = [
    "download-twitter",
    "bench-twitter",
    "download-livejournal",
    "bench-livejournal",
]
```

Comment (mbudiu-vmw) [reply rust/ddlog_benches/Makefile.toml:3]: Is this really necessary? It looks to me like when you run 'cargo make benchmarks' these will be done through transitive dependencies anyway. For me it works fine as it is in this PR.

File: .gitignore (hunk 22 above)
Comment (mbudiu-vmw) [reply .gitignore:34]: Why should it be in .git/info/exclude? I assume no one wants to commit that file by mistake, as I did.

Comment (Kixiron) [reply .gitignore:34]: The file existing isn't part of the project, it's an artifact of your particular dev flow

Comment (mbudiu-vmw) [reply .gitignore:34]: It was generated when running some tests with their default flags, I didn't do anything special for that.

File: rust/template/differential_datalog/src/program/mod.rs [hunk 23]
```diff
@@ -85,7 +89,110 @@ type TKeyAgent<S> = TraceAgent<KeyTrace<S>>;
 type TValEnter<P, T> = TraceEnter<TValAgent<P>, T>;
 type TKeyEnter<P, T> = TraceEnter<TKeyAgent<P>, T>;
 
-/// Diff associated with records in differential dataflow
+#[derive(
```
Comment (ryzhyk) [inline rust/template/differential_datalog/src/program/mod.rs:92]: Do we need `Serialize`/`Deserialize` for weights?  The auto-derived implementations will likely not do what you want here.  I think you also need `#[serde(transparent)]`

File: rust/template/differential_datalog/src/program/mod.rs (hunk 5 above)
Comment (ryzhyk) [reply rust/template/differential_datalog/src/program/mod.rs:134]: I actually find the original version easier to read. @Kixiron , so you think `Self` is more idiomatic?
In any case, @mbudiu-vmw , since you made this change everywhere you probably want it here as well.

File: rust/template/differential_datalog/src/program/mod.rs (hunk 1 above)
Comment (ryzhyk) [reply rust/template/differential_datalog/src/program/mod.rs:146]: Why not use `checked_neg`?

File: rust/template/differential_datalog/src/program/mod.rs (hunk 9 above)
Comment (ryzhyk) [reply rust/template/differential_datalog/src/program/mod.rs:167]: Do you want to apply @Kixiron 's suggestion here and in other places below?

File: rust/template/differential_datalog/src/program/mod.rs (hunk 11 above)
Comment (ryzhyk) [reply rust/template/differential_datalog/src/program/mod.rs:180]: Why do we have this `impl` specifically for `i8`?

File: rust/template/differential_datalog/src/program/mod.rs (hunk 23 above)
Comment (mbudiu-vmw) [reply rust/template/differential_datalog/src/program/mod.rs:92]: In general I added stuff that seemed to be necessary or the program would not compile.
I don't really understand why Abomonation is needed either.
I can try to take it out and see what breaks.
I really don't understand why the auto-derived implementation would be wrong.

File: rust/template/differential_datalog/src/program/mod.rs (hunk 5 above)
Comment (Kixiron) [reply rust/template/differential_datalog/src/program/mod.rs:134]: Yah, `Self` is more idiomatic and it's more resistant to annoyances while refactoring since it just refers to the current struct, one less thing to change

**Comment by mbudiu-vmw** [comment]
There is no checked_neg in num.

**Comment by ryzhyk** [comment]
> There is no checked_neg in num.

Can you use this: https://doc.rust-lang.org/stable/std/primitive.i32.html#method.checked_neg
